src/lib/common/include/sol-atomic.h

```diff
@@ -0,0 +1,229 @@
+ * Returns the previous value state of @p flag and marks it as set, using memory order @p memory_order.
+ */
+
+/**
+ * @fn void sol_atomic_clear(atomic_flag *flag, int memory_order)
+ * Clears the state in the flag @p flag using memory order @p memory_order.
+ */
+
+/**
+ * @typedef sol_atomic_int
+ * An atomic variable compatible with an int.
+ */
+
+/**
+ * @typedef sol_atomic_uint
+ * An atomic variable compatible with an unsigned int.
+ */
+
+/**
+ * @typedef sol_atomic_size_t
+ * An atomic variable compatible with a size_t.
+ */
+
+/**
+ * @typedef sol_atomic_uintptr_t
+ * An atomic variable compatible with an uintptr_t.
+ */
+
+/**
+ * @fn void sol_atomic_store(atomic *object, type value, int memory_order)
+ * Stores value @p value at @p object, using the memory order @p memory_order.
+ */
+
+/**
+ * @fn value sol_atomic_load(atomic *object, int memory_order)
+ * Loads a value from @p object using memory order @p memory_order and returns it.
+ */
+
+/**
+ * @fn value sol_atomic_exchange(atomic *object, value new_value, int memory_order)
+ * Atomically replaces the value stored at @p object with new_value, using memory order
+ * @p memory_order, and returns the old value.
+ */
+
+/**
+ * @fn bool sol_atomic_compare_exchange(atomic *object, value *expected, value desired, int memory_order_success, int memory_order_failure)
+ * Executes an atomic compare-and-swap operation at @p object: if the value stored
+ * there is @p expected, then this function replaces it with @p desired, using memory
+ * order @p memory_order_success and returns @c true. If the value at @p object
+ * is not @p desired, then this function loads the current value into @p expected using
+ * memory order @p memory_order_failure and returns @c false.
+ *
+ * This function is equivalent to C11's strong compare_exchange and its requirements on
+ * the memory ordering arguments apply here too.
+ */
+
+/**
+ * @fn value sol_atomic_fetch_add(atomic *object, value addend, int memory_order)
+ * Atomically adds @p added to @p object and returns the new value, using memory order @p memory_order.
+ */
+
+#if (defined(__STDC_VERSION__) && __STDC_VERSION__ > 201112L) \
+    || (defined(__GNUC__) && !defined(__clang__) && (__GNUC__ * 100 + __GNUC_MINOR__ > 408)) \
+    || (defined(__clang__) && (__clang_major__ * 100 + __clang_minor__ >= 307)) \
+    || defined(DOXYGEN_RUN)
+/* GCC 4.9, Clang 3.7 provide <stdatomic.h>, a C11 header */
+#include <stdatomic.h>
+
+/* Initialization */
+#define SOL_ATOMIC_INIT         ATOMIC_VAR_INIT
+#define SOL_ATOMIC_FLAG_INIT    ATOMIC_FLAG_INIT
+
+/* Order levels */
+#define SOL_ATOMIC_RELAXED memory_order_relaxed
+#define SOL_ATOMIC_CONSUME memory_order_consume
+#define SOL_ATOMIC_ACQUIRE memory_order_acquire
+#define SOL_ATOMIC_RELEASE memory_order_release
+#define SOL_ATOMIC_ACQ_REL memory_order_acq_rel
+#define SOL_ATOMIC_SEQ_CST memory_order_seq_cst
+
+/* Types */
+typedef atomic_int sol_atomic_int;
```

Comment:
We got atomic_uchar but I guess it's no big deal to fallback to int on boolean atomics.